docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+<br/>
+
+##### Upgrades providers and changing contract
+
+If the new version of the provider does abide by a new version of the Cluster
+API contract, it is required to ensure all the other providers in the
+management cluster should get the new version too.
+
+![Figure 2](./images/management-cluster-operator/fig2.png "Figure for
+upgrading provider and changing contract")
+<div align="center">Upgrading providers and changing contract</div>
+<br/>
+
+As a first step, it is required to pause all the providers by setting the
+`spec.Paused` field to true for each provider; the operator will block any
+contract upgrade until all the providers are paused.
+
+After all the providers are in paused state, you can proceed with the upgrade
+as described in the previous paragraph (change the `spec.Version` field).
+
+Once all the providers are upgraded to a version that abides to the new
+contract, it is possible for the operator to unpause providers; the operator
+does not allow to unpause providers if there are still providers abiding to
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
+
+##### Deleting a provider
+
+In order to delete a provider you have to delete the corresponding provider
+object.
+
+Deletion of the provider will be blocked if any workload cluster using the
+provider still exists.
+
+Additionally, deletion of a core provider should be blocked if there are still
+other providers in the management cluster.
+
+#### Operator Lifecycle Management
+
+##### Operator Installation
+
+- `clusterctl init` will install the operator and its corresponding CRDs as a
+  pre-requisite if the operator doesn’t already exist. Please note that this
+  command will consider image overrides defined in the local clusterctl config
+  file.
+- If the admin does not want to use clusterctl to install the operator, it is
+  possible to kubectl apply the operator yaml that will be published in the
+  cluster-api release artifacts.
+- There can be only one operator installed per management cluster, across all
+  namespaces.
+
+##### Operator Upgrade
+
+- The admin can use `clusterctl upgrade operator` to upgrade the operator
+  components. Please note that this command will consider image overrides
+  defined in the local clusterctl config file.
```

Comment:
I think we should make this also part of the `clusterctl upgrade plan` and part of the `clusterctl upgrade apply` workflow similarly to what we are doing for cert-manager.
This does not exclude the idea for providing separated commands for upgrading the operator alone (or cert-manager), but I would like clusterctl to always provide an easy button for all the operations.

Reply:
Make sense. Since we currently check to upgrade cert-manager as part of `upgrade plan` and `upgrade apply` we can do the same for the operator.